metadata-ingestion/src/datahub/ingestion/source/mongodb.py

```diff
@@ -0,0 +1,121 @@
+    options: dict = {}
+
+    database_pattern: AllowDenyPattern = AllowDenyPattern.allow_all()
+    collection_pattern: AllowDenyPattern = AllowDenyPattern.allow_all()
+
+
+@dataclass
+class MongoDBSourceReport(SourceReport):
+    filtered: List[str] = field(default_factory=list)
+
+    def report_dropped(self, name: str) -> None:
+        self.filtered.append(name)
+
+
+@dataclass
+class MongoDBSource(Source):
+    config: MongoDBConfig
+    report: MongoDBSourceReport
+
+    def __init__(self, ctx: PipelineContext, config: MongoDBConfig):
+        super().__init__(ctx)
+        self.config = config
+        self.report = MongoDBSourceReport()
+
+        options = {}
+        if self.config.username is not None:
+            options["username"] = self.config.username
+        if self.config.password is not None:
+            options["password"] = self.config.password
+        if self.config.authMechanism is not None:
+            options["authMechanism"] = self.config.authMechanism
+        options = {
+            **options,
+            **self.config.options,
+        }
+
+        self.mongo_client = pymongo.MongoClient(self.config.connect_uri, **options)
+
+        # This cheaply tests the connection. For details, see
+        # https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient
+        self.mongo_client.admin.command("ismaster")
+
+    @classmethod
+    def create(cls, config_dict: dict, ctx: PipelineContext):
+        config = MongoDBConfig.parse_obj(config_dict)
+        return cls(ctx, config)
+
+    def get_workunits(self) -> Iterable[MetadataWorkUnit]:
+        env = "PROD"
+        platform = "mongodb"
+
+        database_names: List[str] = self.mongo_client.list_database_names()
+        for database_name in database_names:
+            if database_name in DENY_DATABASE_LIST:
+                continue
+            if not self.config.database_pattern.allowed(database_name):
+                self.report.report_dropped(database_name)
+                continue
+
+            database = self.mongo_client[database_name]
+            collection_names: List[str] = database.list_collection_names()
+            for collection_name in collection_names:
+                dataset_name = f"{database_name}.{collection_name}"
+                if not self.config.collection_pattern.allowed(dataset_name):
+                    self.report.report_dropped(dataset_name)
+                    continue
+
+                mce = MetadataChangeEvent()
+                dataset_snapshot = DatasetSnapshot()
+                dataset_snapshot.urn = f"urn:li:dataset:(urn:li:dataPlatform:{platform},{dataset_name},{env})"
+
+                dataset_properties = DatasetPropertiesClass(
+                    tags=[],
+                    customProperties={},
+                )
+                dataset_snapshot.aspects.append(dataset_properties)
+
+                # TODO: Guess the schema via sampling
```

Comment:
Would be great if mongo provided this type of capability for us.. ie tracking the superset document

Reply:
There's nothing for it in mongodb itself, although the MongoDB Compass app does have this, albeit written in JS